Ruleset: AsyncAPI

nulltoken opened the pull request:
Fixes #965
Fixes #789

**Checklist**

- [ ] Tests added / updated
- [ ] Docs added / updated
- [ ] Log issue to add a rule to ensure all variables used in server url are defined
- [ ] Log issue to add a rule to ensure all defined server variables are used
- [ ] Log issue to add a rule to ensure all parameters used in channel path are defined
- [ ] Log issue to add a rule to ensure all defined channel parameters are used
- [ ] Log issue to add a rule to identify orphaned components (messageTraits, ...)
- [ ] Migrate `asyncApi2PayloadValidation` to a custom function of the ruleset

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

> If indicated yes above, please describe the breaking change(s).
>
> **Remove this quote before creating the PR.**

**Screenshots**

If applicable, add screenshots or gifs to help demonstrate the changes. If not applicable, remove this screenshots section before creating the PR.

**Additional context**

Add any other context about the pull request here. Remove this section if there is no additional context.

nulltoken commented on scripts/generate-assets.js:
```diff
@@ -29,32 +29,46 @@ if (!fs.existsSync(baseDir)) {
+        })
+        console.log('awaited ' + target)
+        content = JSON.stringify((x).result);
+        console.log('done ' + target);
```
@P0lip I might need help here as I'm hitting a wall. `generate-assets` eventually makes the build timeout.

My lame attempt at trying to understand the issue led me to this.

![image](https://user-images.githubusercontent.com/92363/75291206-1143a680-5822-11ea-9ab3-f9cfdc095601.png)

The resolver seems to hit some infinite loop while processing the json schema draft 7 (that asyncapi depends on)

![image](https://user-images.githubusercontent.com/92363/75291535-b2326180-5822-11ea-86ae-4bde59ca1f95.png)

Any idea? (please, say "Yes").

P0lip replied:
Urgh, I think I know what the issue is. This is something Phil faced some time ago, see
https://github.com/stoplightio/spectral/issues/789. I believe the most reasonable solution would be to use json-schema-ref-parser for dereferencing

nulltoken replied:
@P0lip On it. Thanks! <3

nulltoken replied:
@P0lip Thanks to you, I went a step farther. However, I need to get a better understanding of the overall constraints governing the use of those static assets:

 - Are they only used in Spectral test environments?
 - Are there some parts of the Studio ecosystem that leverage the `Spectral.registerStaticAssets()` method?

nulltoken replied:
@P0lip @philsturgeon 🎉 At last! I got it. Took some time.

`json-schema-ref-parser` ended up being a dead-end because of an issue I've hit. I'll open an issue upstream.

I had to slightly "hack" the vendored aas2 schema. This allowed me to reuse most of the current code with minimal changes (I had to inline this **[part](https://github.com/asyncapi/asyncapi/blob/master/versions/2.0.0/schema.json#L358)**).

nulltoken commented:
@philsturgeon Few questions:
- In the original document, some rules were commented out. I didn't consider them. Should I try and re-integrate them or were they left out on purpose?
- Some rules are identical between oas and aas. Should this stay as is or should we try to somehow reuse a common set of shared rules between the two rulesets?

@P0lip @philsturgeon Next steps:
- I believe I've now got the overall skeleton in place (@P0lip Would you have some time for a first review pass?). I'm now going to create tests for each of those new rules (at least one passing and one failing per rule).
- Then some documentation for the new ruleset (@philsturgeon I might need your help on that)

philsturgeon commented on rollup.config.js:
```diff
@@ -6,18 +6,27 @@ import commonjs from 'rollup-plugin-commonjs';
-
-  functions.push(targetFile);
+
+const builtIns = ['oas', 'aas']
```
We're kinda stuck with `oas` due to it being chosen a while back and not being worthwhile to rename it just yet, but for new things lets call it a name which people will recognize. I don't think aas is a thing, so let's just call it `asyncapi`.

```suggestion
const builtIns = ['oas', 'asyncapi']
```

philsturgeon commented:
@nulltoken the commented out ones were things I didn't delete or make work. I deleted some things which were definitely not worth trying to port over, and left in some things I hadn't researched. If those rules are relevant lets get them working, if they are not we can delete them. 

> should we try to somehow reuse a common set of shared rules between the two rulesets?

A little bit of duplication is ok for now, I'd rather get it working than spend a bunch of time on DRYing it prematurely.

philsturgeon commented:
@nulltoken the commented out ones were things I didn't delete or make work. I deleted some things which were definitely not worth trying to port over, and left in some things I hadn't researched. If those rules are relevant lets get them working, if they are not we can delete them. 

> should we try to somehow reuse a common set of shared rules between the two rulesets?

A little bit of duplication is ok for now, I'd rather get it working than spend a bunch of time on DRYing it prematurely.

nulltoken commented:
Dear journal,

At last! Back in the world of CI green. Pretty relieved! 

Suffered a lot in Karma land and weird execution of jest BeforeAll() test hooks. Had to switch over to BeforeEach(). A bit of a waste as we keep on rebuilding something which is supposed to be immutable, but, at least... it works.

nulltoken commented:
Dear Diary,

I've started to work back on the meat of this PR: checking the rules. By doing so, I've hit yet another issue related to jest test hooks and mocking of node modules. A helper function, hosted in `setupTest.ts` triggers too soon and prevents some tests other tests to successfully mock the 'nanoid' module. After some head banging, I've opted to get rid of the mocking, while trying to keep the intent of the (slightly) altered tests:

- Failing build (before the removal of nanoid mocking): https://circleci.com/gh/stoplightio/spectral/11952
- Passing build (after the removal of nanoid mocking): https://circleci.com/gh/stoplightio/spectral/11953

@P0lip Could you please take a look at the `chore: avoid mocking nanoid node module` commit and see if that option fits you?

P0lip commented:
@nulltoken as chatted on Slack, https://github.com/stoplightio/spectral/pull/1025 should fix the issue.

nulltoken commented:
@philsturgeon Won't eventually port `example-value-or-externalValue` as this seems to be an OAS-only thingie, unsupported by AsyncApi.

philsturgeon commented:
Great! Delete anything that’s not useful

On Wed, Apr 8, 2020 at 09:32, nulltoken <notifications@github.com> wrote:

> [@philsturgeon](https://github.com/philsturgeon) Won't eventually port example-value-or-externalValue as this seems to be an OAS-only thingie, unsupported by AsyncApi.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/stoplightio/spectral/pull/974#issuecomment-610827294), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAAQONK3OTPGTNP4VOXBAMDRLQZC7ANCNFSM4KXS7NZA).

nulltoken commented:
@P0lip @philsturgeon This PR is coming to an end. At last.

Although it's not finished yet, I think now would be a good time to start taking a look at the proposed changes.

@philsturgeon Could you please take a look at the ruleset and ensure
 - the rules names are ok
 - the descriptions/messages are explicit enough
 - the proposed severity fits you
 - some of the rules have been extracted out from the oas rulesets and bear some `tags`. I have no idea what is their intended purpose. If they're used in a different part of the ecosystem, could you please either propose changes with the proper tags for the rules or explain me what they mean, the list of valid values and what each value refer to? If they're only remnants of a distant past and should be dropped, I'll be happy to do so. FWIW, I haven't found any information about those in the doco.

Once we're all good, I'll document those rules in the `asyncapi-rules.md` doc and will bother you again for a review.

I've also added/updated some `.md` files, could you please take a peek at them?

@P0lip I'm sorry for the lot of lines of code I expect you to review. I've extracted as much as possible in seperate PRs, but still, a lot of them remain.

The tests should be pretty boring to read, but they're the real meat of this PR. I've tried to be as careful as possible when writing them and I _hope_ all of them pass for good reasons (but, there's still a chance, I made a mistake and that both the test and the code are wrong :-( ).

I've tried to write as terse (and performant) JsonPath queries as possible. For instance, I've opted to only validate `default`, `example` and `examples` at the first level of each schema definition (Would that not be acceptable, could we postpone the extension to a follow-up PR?).

⚠️ The `default`, `example` and `examples` leverages Json Schema Draft 7 and _not_ the AsyncApi2 schema object (as it should be done according to the spec). I'm working on a little something to make it happen.

nulltoken commented:
> ⚠️ The `default`, `example` and `examples` leverages Json Schema Draft 7 and _not_ the AsyncApi2 schema object (as it should be done according to the spec). I'm working on a little something to make it happen.

Woot! I made it work. Now let's see if I can package it into something usable.

P0lip commented on asyncapi.yaml:
```diff
@@ -0,0 +1,351 @@
+#From https://github.com/openapi-contrib/style-guides/pull/4/files
+# These rules will be helping the AsyncAPI people write be better quality, but 
+# have no thoughts on the actual contents of the API
+
```
Could we get rid of these empty lines?
It's likely just me, but I find them distracting and they kind of make the processing of document more difficult. :sweat_smile:

nulltoken replied:
@P0lip That file was only a todolist to me. I dropped it.

P0lip commented on src/rulesets/reader.ts:
```diff
@@ -65,15 +65,17 @@ const createRulesetProcessor = (
-            // happens
-          }
-          return opts;
+          return new Promise<IUriParserResult>((resolve, reject) => {
```
Is there any particular reason we need this promise here?
We have `async` fn in place already.

nulltoken replied:
Duh. :-/
...
Fixed. ;-)

P0lip commented on README.md:
```diff
@@ -4,14 +4,16 @@
 [![Treeware (Trees)](https://img.shields.io/treeware/trees/stoplightio/spectral)](https://plant.treeware.earth/stoplightio/spectral)
 
-A flexible JSON/YAML linter, with out of the box support for OpenAPI v2 and v3.
+A flexible JSON/YAML linter, with out of the box support for OpenAPI v2/v3 and AsyncAPI v2.
```
We gotta finally migrate all these oas/aas specs away to a separate repo. :sweat_smile:

philsturgeon replied:
Yep this was always going to happen eventually and I think v6.0 would be perfect.

1. Documentation on how to include a ruleset via NPM
2. In v6.0 we split these out into their own modules which are then automatically included, in v6.1 we emit warnings suggesting folks should include them directly
3. Spectral v7.0 they're no longer included by default

Dunno if you could sneak number 1 in between your million other tasks? Take a stab at getting all the information in dont worry about wording, I'll polish it up.

P0lip replied:
> Dunno if you could sneak number 1 in between your million other tasks? Take a stab at getting all the information in dont worry about wording, I'll polish it up.

Yeah, I've got a sample repo with an example ruleset up and running, so can simply copy & paste it.
Will do it tomorrow.

P0lip commented on asyncapi.yaml:
```diff
@@ -0,0 +1,351 @@
+#From https://github.com/openapi-contrib/style-guides/pull/4/files
```
Actually, I feel like this file is supposed to be removed?

philsturgeon replied:
Yeah this is gonna go away so lets not mention it.

nulltoken replied:
@philsturgeon @P0lip That file was only a todolist to me. I dropped it.

philsturgeon commented:
ping @fmvilas @jerzyn here we are! Can you check this out?

fmvilas commented on README.md:
```diff
@@ -4,14 +4,16 @@
 
 - Create custom rules to lint JSON or YAML objects
-- Ready to use rules to validate and lint OpenAPI v2 _and_ v3 documents
+- Ready to use rules to validate and lint
```
```suggestion
- Ready-to-use rules to validate and lint:
```

nulltoken replied:
@fmvilas Fixed

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1 @@
+# AsyncAPI Rules
```
Is this file incomplete?

nulltoken replied:
@fmvilas Before diving into the documentation part of the PR, I was expecting some comments from @philsturgeon with regards to the ruleset first.

This file will eventually list all the available rules.

nulltoken replied:
@fmvilas @philsturgeon The file now list all the proposed rules. Could you please take a look at them?

Every re-wording proposal would be greatly appreciated! ;-)

fmvilas commented on docs/getting-started/asyncapi.md:
```diff
@@ -0,0 +1,5 @@
+
+Spectral is a generic linter, but a lot of effort has been put in to making sure [AsyncAPI v2](https://www.asyncapi.com/docs/specifications/2.0.0/) is well supported.
+
+Run Spectral against a document without specifying a ruleset will trigger an auto-detect, where Spectral will look to see if `asyncapi: 2.0.0` is at the top of the file. If it finds it, it will load `spectral:asyncapi`, which is documented in our [Reference > AsyncAPI Rules](../reference/asyncapi-rules.md).
```
> ...where Spectral will look to see if `asyncapi: 2.0.0` is at the top of the file

Is it looking for `asyncapi: 2.0.0` to be at the top of the file or the root of the YAML document? I'm asking because it may not be the first line.

nulltoken replied:
@fmvilas the engines searches for a node named `asyncapi` at the root of the document, then ensures it's valued with `2.0.0`.

@philsturgeon How should we reword that so that it's clearer?

philsturgeon replied:
"
```suggestion
Run Spectral against a document without specifying a ruleset will trigger an auto-detect, where Spectral will look to see if `asyncapi: 2.0.0` in the root of the document. If it finds it, it will load `spectral:asyncapi`, which is documented in our [Reference > AsyncAPI Rules](../reference/asyncapi-rules.md).
```

nulltoken replied:
@philsturgeon @fmvilas Fixed

nulltoken commented:
@fmvilas @derberg Thanks for the feedback. Unfortunaley, it looks like I led you (by mistake) on a wrong track. The file named `asyncapi.yml` was only some kind of todolist to me I used to keep track of the remaining rules to implement.

The real "meat" of this PR is
- the ruleset (located at `/src/rulesets/asyncapi/index.json`)
- the tests (located at `/src/rulesets/asyncapi/__tests__`) exercizing this ruleset that should better help you understand how the rules behave

Would you agree to spend a little time taking another look at this PR, I'd be especially grateful.

⚠️ Full disclosure: I've never written any AnyAPI spec, so my understanding of it is very thin. I've done my best _trying_ to respect the spec as much as possible, but there are great chances I misunderstood part of it.

nulltoken commented on src/rulesets/asyncapi/schemas/schema.asyncapi2.json:
```diff
@@ -0,0 +1,1599 @@
+{
+  "title": "AsyncAPI 2.0.0 schema.",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "asyncapi2",
```
@fmvilas Here I "hacked" the initial schema by adding an `$id`...

nulltoken commented on src/rulesets/asyncapi/schemas/schema.asyncapi2.json:
```diff
@@ -0,0 +1,1599 @@
+      "description": "JSON objects describing re-usable channel parameters."
+    },
+    "schema": {
+      "$id": "#schemaObject",
```
@fmvilas and another `$id` here. 

In the spec, the `schemaFormat` is defined as _"A string containing the name of the schema format used to define the message payload. If omitted, implementations should parse the payload as a Schema object. Check out the supported schema formats table for more information. Custom values are allowed but their implementation is OPTIONAL. A custom value MUST NOT refer to one of the schema formats listed in the table."_

The way I interpreted it is that when `schemaFormat` is `undefined` then the payload should be valid against `#/definitions/schema`. 

Similarly, any `example` (or `default`) of that payload should be also valid against both the the `payload` schema and `#/definitions/schema`.

As such, I decorated both the AsyncApi2 schema and the Schema Object with a JsonSchema $id to be able to look them up more easily.

Would have I properly understood the spec, would you consider updating the upstream spec with some `$id` as well? That might make the job of other validators a bit easier.

nulltoken replied:
@fmvilas Hmmm. I got that wrong. Only the payload schema should be valid against the SchemaObject. `default` and `example`, `examples` should be valid against the payload schema only. I'll fix that in a following commit.

nulltoken replied:
@fmvilas The valid-payload-* rules and tests have been fixed.

fmvilas replied:
Awesome! I created an issue so we don't forget: https://github.com/asyncapi/asyncapi/issues/373. Thanks!

philsturgeon commented on src/formats/asyncapi.ts:
```diff
@@ -0,0 +1,10 @@
+};
+
+export const isAsyncApiv2 = (document: unknown) =>
+  bearsAStringPropertyNamed(document, 'asyncapi') && String((document as MaybeAsyncApi2).asyncapi) === '2.0.0';
```
Much like OpenAPI, the patch version should be ignored. 

https://github.com/stoplightio/spectral/blob/develop/src/formats/openapi.ts

```
export const isOpenApiv3 = (document: unknown) =>
  isObject(document) && 'openapi' in document && parseFloat(String((document as MaybeOAS3).openapi)) === 3;
```

Notice how it's just looking for the major "3" and not worried about the minor or patch?

nulltoken replied:
@philsturgeon Fixed

Will now accept `2.{minor}.{patch}` format where

`minor` or `patch`:
 - can only be made of numbers
 - can not start with a `0`

philsturgeon commented on src/rulesets/asyncapi/__tests__/asyncapi2-channel-path-no-empty-parameter-declaration.ts:
```diff
@@ -0,0 +1,45 @@
+import { Spectral } from '../../../spectral';
+import { IRunRule } from '../../../types';
+
+const ruleName = 'asyncapi2-channel-path-no-empty-parameter-declaration';
```
ooof can we find a shorter name for this? We might trigger some UI issues like this.

nulltoken replied:
@philsturgeon Ok. I'll try and come up with shorter names.

nulltoken replied:
@philsturgeon Shorter names! Could you please check they still make sense?

nulltoken commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,465 @@
+        }
+      },
+      "tags": [
+        "given"
```
@philsturgeon What are those tags for? Do we need them?

philsturgeon replied:
I think tags are an artifact which was going to be used for something, but never was. @P0lip can you confirm? If so, yeah lets kill off the tags.

P0lip replied:
I don't know what the purpose of tags is. They had been added added before I started to work on spectral.

nulltoken replied:
@philsturgeon So 💀 them with 🔥 ?

philsturgeon replied:
💣💥

nulltoken replied:
They're gone, now.

nulltoken commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,465 @@
+    "asyncapi2-headers-schema-type-object": {
+      "description": "Headers schema type should be `object`.",
+      "message": "Headers schema type should be `object` ({{error}}).",
+      "severity": 0,
```
@philsturgeon Could you please check this severity (and the other ones that are defined) make sense?

philsturgeon replied:
I forget what the numbers mean, I think we can use "warn" and "error" right @P0lip

nulltoken replied:
@philsturgeon Fixed

nulltoken commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,465 @@
+    },
+    "asyncapi2-info-contact-properties": {
+      "description": "Contact object should have `name`, `url` and `email`.",
+      "recommended": true,
```
@philsturgeon Could you please also check if the `recommended` values make sense as well?

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+### asyncapi2-channel-no-empty-parameter
+
+Path parameter declarations cannot be empty, ex.`/given/{}` is invalid.
```
```suggestion
Channel parameter declarations cannot be empty, ex.`/given/{}` is invalid.
```

nulltoken replied:
@fmvilas Fixed

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+### asyncapi2-channel-no-query-nor-fragment
+
+Query parameters and fragments shouldn't be used in paths. Instead use bindings to define them.
```
```suggestion
Query parameters and fragments shouldn't be used in channel names. Instead, use bindings to define them.
```

nulltoken replied:
@fmvilas Fixed

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+### asyncapi2-channel-no-trailing-slash
+
+Keep trailing slashes off of paths, as it can cause some confusion. Some web tooling (like mock servers, real servers, code generators, application frameworks, etc.) will treat `example.com/foo` and `example.com/foo/` as the same thing, but other tooling will not. Avoid any confusion by just documenting them without the slash, and maybe some tooling will let people shove a / on there when they're ...
```
```suggestion
Keep trailing slashes off of channel names, as it can cause some confusion. Most messaging protocols will treat `example/foo` and `example/foo/` as different things. Keep in mind that tooling may replace slashes (`/`) with protocol-specific notation (e.g., `.` for AMQP), therefore, a trailing slash may result in an invalid channel name in some protocols.
```

nulltoken replied:
@fmvilas Fixed

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+### asyncapi2-operation-operationId
+
+This operation ID is essentially a reference for the operation, which can be used to visually suggest a connection to other operations. This is like some theoretical static HATEOAS-style referencing, but it's also used for the URL in some documentation systems.
```
I think this description is very http-ish. There's no HATEOAS or URLs in most of the messaging systems.

```suggestion
This operation ID is essentially a reference for the operation. Tools may use it for defining function names, class method names, and even URL hashes in documentation systems.
```

nulltoken replied:
@fmvilas Fixed

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+This operation ID is essentially a reference for the operation, which can be used to visually suggest a connection to other operations. This is like some theoretical static HATEOAS-style referencing, but it's also used for the URL in some documentation systems.
+
+Make the value `lower-hyphen-case`, and try and think of a name for the action which does not relate to the HTTP message. Base it off the actual action being performed. `create-polygon`? `search-by-polygon`? `filter-companies`?
```
Why this recommendation? I'd say that camelCase is the one that works the best in most places. Most of the times the operationId is used for function or method names and URL hashes. `lower-hyphen-case` is usually not a friendly notation in programming languages and URL hashes also support camelCase.

nulltoken replied:
@fmvilas Dropped.

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+### asyncapi2-payload-example
+
+`example` objects should be valid against the `payload` they decorate.
```
This one doesn't exist on AsyncAPI. Only `examples`.

nulltoken replied:
@fmvilas Dropped.

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+### asyncapi2-schema-example
+
+`example` objects should be valid against the `schema` they decorate.
```
Same as above. This one doesn't exist on AsyncAPI.

nulltoken replied:
@fmvilas Dropped.

fmvilas commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,325 @@
+
+Server URL should not have a trailing slash.
+
+Some tooling forgets to strip trailing slashes off when it's joining the `servers.url` with `channels`, and you can get awkward URLs like `https://example.com/api//pets`. Best to just strip them off yourself.
```
```suggestion
Some tooling forgets to strip trailing slashes off when it's joining the `servers.url` with `channels`, and you can get awkward URLs like `https://example.com/api//pets`. Best to just strip them off yourself.
```

nulltoken replied:
@fmvilas Fixed (switched to mqtt protocol)

derberg commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,435 @@
+      }
+    },
+    "asyncapi2-headers-schema-type-object": {
+      "description": "Headers schema type should be `object`.",
```
this is a specification requirement (https://www.asyncapi.com/docs/specifications/2.0.0/#a-name-messageobject-a-message-object), the parser will find out about this issue. Are those also parts that spectral ruleset should check?

nulltoken replied:
@derberg I may need some help on this topic:

Given the following document

```
  {
      "asyncapi": "2.0.0",
      "info": {
          "title": "Valid AsyncApi document",
          "version": "1.0"
      },
      "channels": {},
      "components": {
          "messages": {
              "userSignUp": {
                  "summary": "Action to sign a user up.",
                  "description": "Multiline description of what this action does.\nHere you have another line.\n",
                  "tags": [{
                          "name": "user"
                      },
                      {
                          "name": "signup"
                      }
                  ],
                  "headers": {
                      "type": "integer",
                      "enum": [1, 2]
                  },
                  "payload": {
                      "type": "integer"
                  }
              }
          }
      }
  }
```

The asyncapi2 schema doesn't seem to catch the wrong `headers` type.

Help?

derberg replied:
@nulltoken I would say it is a bug in our parser that doesn't validate header, spec is clear here, should be `object`

derberg replied:
I reported a bug for us https://github.com/asyncapi/parser-js/issues/68

nulltoken replied:
@derberg Awesome! I subscribed to the issue. When it's closed, I'll drop this rule.

fmvilas commented on docs/reference/asyncapi-rules.md:
````diff
@@ -0,0 +1,325 @@
+
+``` yaml
+servers:
+  - url: https://example.com
+  - url: https://example.com/api
````
It would be great not to use HTTP as it's one of the less common protocols in messaging.

```suggestion
  - url: mqtt://example.com
  - url: mqtt://example.com/broker
```

nulltoken replied:
@fmvilas Fixed

fmvilas commented on docs/reference/asyncapi-rules.md:
````diff
@@ -0,0 +1,325 @@
+
+``` yaml
+servers:
+  - url: https://example.com/
+  - url: https://example.com/api/
````
```suggestion
  - url: mqtt://example.com/
  - url: mqtt://example.com/broker/
```

nulltoken replied:
@fmvilas Fixed

derberg commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,435 @@
+      }
+    },
+    "asyncapi2-server-no-empty-variable": {
+      "description": "Server URL should not have empty variable substitution pattern.",
```
the awesome rule would be that checks that you have a variable in your server url {port} but you do not describe it with variable object -> https://www.asyncapi.com/docs/specifications/2.0.0/#serverVariableObject which make the server url useless

derberg replied:
another awesome valid rule is for the same use case, for variables but in channel names

nulltoken replied:
@deberg This is on my wishlist as well. This kind of rule would require a custom function. I've updated the description of that PR to keep track of these rules and add them through another PR. If you can think of other rules that would make sense, please share them!

philsturgeon commented:
@fmvilas thank you! I would have caught more of these HTTP bits, I was just having my coffee. 😅 

We'll get these HTTP specific bits trimmed out and tidied up soon.

fmvilas commented:
No problem, mate! I'm loving the work you folks are doing :)

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,60 @@
+import * as AJV from 'ajv';
```
Whoa, I'd rather not do this, since it'll inline the whole ajv, significantly increasing the size of a function.

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,60 @@
+import * as AJV from 'ajv';
+import * as jsonSpecv7 from 'ajv/lib/refs/json-schema-draft-07.json';
```
Same. We need to find some way to avoid doing this.
I think I have an idea on how to expose it. Will open a PR today.

P0lip replied:
@nulltoken https://github.com/stoplightio/spectral/pull/1100

nulltoken replied:
@P0lip #1100 is merged and this PR has been rebased on top of it in order to leverage it.

Sadly, I'm still unable to migrate this to a custom function (as rollup cringes about tslib). Proposal is to keep it as a built-in function and migrate it to a custom function once this is merged. 

I've updated the description of this PR to keep track of this through an upcoming issue.

P0lip commented on src/functions/schema.ts:
```diff
@@ -155,46 +159,63 @@ export const schema: IFunction<ISchemaOptions> = (targetVal, opts, paths) => {
     }
   }
 
+  PerformSchemaValidation(validator, targetVal, results, schemaObj, path, opts.prepareResults);
```
This must go in the try block.

nulltoken replied:
@P0lip Fixed.

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,60 @@
+  // The subsection of the targetVal which contains the good bit
+  const relevantItems = getLintTargets(targetVal, opts.field);
+
+  const rootPath = [...(paths.target || paths.given)];
```
```suggestion
  const rootPath = paths.target || paths.given;
```

nulltoken commented:
> I went through `index.json` and added few comments.
> One open question is: there are some rules without type, is it intentional?

@derberg Fixed.

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,311 @@
+
+In your ruleset file you can add `extends: "spectral:asyncapi"` and you'll get all of the following rules applied.
+
+## AsyncAPI v2-only
```
```suggestion
```

we only had this in here when we were documenting both oas2 and oas3 at the same time.

nulltoken replied:
@philsturgeon Fixed

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,311 @@
+
+These rules will only apply to AsyncAPI v2 documents.
+
+### asyncapi2-channel-no-empty-parameter
```
We only had version numbers in the rules when a rule only applied to a specific version. We'll never support `asyncapi1` and we dont know what 3 will look like, these rules might apply and we'd have to rename them later. I think we'd be better off naming them all `asyncapi-*` now to avoid breaking later.

nulltoken replied:
@philsturgeon Fixed

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,311 @@
+
+### asyncapi2-channel-no-query-nor-fragment
+
+Query parameters and fragments shouldn't be used in channel names. Instead, use bindings to define them.
```
Can we get some examples?

nulltoken replied:
@philsturgeon I'm afraid I won't be able to come up with a good/bad example that would make sense in an AsyncAPI context.

@fmvilas Could you please help me here? What's the known bad practices regarding query and fragments and how should they be converted to bindings?

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,311 @@
+
+### asyncapi2-channel-no-trailing-slash
+
+Keep trailing slashes off of channel names, as it can cause some confusion. Most messaging protocols will treat `example/foo` and `example/foo/` as different things. Keep in mind that tooling may replace slashes (`/`) with protocol-specific notation (e.g., `.` for AMQP), therefore, a trailing slash may result in an invalid channel name in some protocols.
```
```suggestion
Keep trailing slashes off of channel names, as it can cause some confusion. Most messaging protocols will treat `example/foo` and `example/foo/` as different things. Keep in mind that tooling may replace slashes (`/`) with protocol-specific notation (e.g.: `.` for AMQP), therefore, a trailing slash may result in an invalid channel name in some protocols.
```

nulltoken replied:
@philsturgeon Fixed

nulltoken replied:
@philsturgeon Fixed

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,311 @@
+
+### asyncapi2-info-contact-properties
+
+The [info-contact](#asyncapi2-info-contact) rule will ask you to put in a contact object, and this rule will make sure it's full of the most useful properties: `name`, `url` and `email`.
```
```suggestion
The [asyncapi-info-contact](#asyncapi-info-contact) rule will ask you to put in a contact object, and this rule will make sure it's full of the most useful properties: `name`, `url` and `email`.
```
these should match

nulltoken replied:
@philsturgeon Fixed

nulltoken replied:
@philsturgeon Fixed

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,311 @@
+
+### asyncapi2-payload-examples
+
+Values of the `examples` array should be valid against the `payload` they decorate.
```
examples here and the default rule above too please, its super valuable and hard to explain.

nulltoken replied:
@philsturgeon Fixed

nulltoken replied:
@philsturgeon Could you please double check the proposed examples are clear enough?

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,311 @@
+
+Potential unused reusable `schema` entry has been detected.
+
+_Warning:_ This rule may identify false positives when linting a specification
```
```suggestion
<!-- theme: warning -->
*Warning:* This rule may identify false positives when linting a specification
```

nulltoken replied:
@philsturgeon Fixed.

philsturgeon commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,398 @@
+    "asyncapi2-info-contact": {
+      "description": "Info object should contain `contact` object.",
+      "recommended": true,
+      "type": "style",
```
@P0lip I've been wondering for a while if we are misusing "type": "style". I think a style recommendation is "use camelCase" and other stylistic only things. Wanting contact information seems more functional than just style. 

Did we ever have a particular definition for when something is style or not, and do we want to figure it out before we bake it into async too?

philsturgeon commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,398 @@
+    },
+    "asyncapi2-payload-unsupported-schemaFormat": {
+      "description": "Message schema validation is only supported with default unspecified `schemaFormat`.",
+      "severity": 2,
```
let's use "warn" "error" "info" for things so folks (incouding me) dont need to work it out? :D

nulltoken replied:
@philsturgeon Fixed.

I've tried to do the same for oas (#1099), but it'll require some more work.
I'll table #1099 until this one is merged.

philsturgeon commented on src/rulesets/asyncapi/index.json:
```diff
@@ -0,0 +1,398 @@
+      "recommended": true,
+      "type": "validation",
+      "given": [
+        "$.components.schemas.*.default^",
```
Is the ^ a JSON Path feature I'm not familiar with?

nulltoken replied:
@philsturgeon It's part of what jsonpath-plus brings:
- `^` to return the parent.
- `*~` to return the keys

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,368 @@
+
+### asyncapi-info-description
+
+OpenAPI object info `description` must be present and non-empty string.
```
Oops! 
```suggestion
AsyncAPI object info `description` must be present and non-empty string.
```

nulltoken replied:
@philsturgeon Fixed

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,368 @@
+  description: >+
+    Some description about the general point of this API, and why it exists when another similar but different API also exists.
+
+    ## Authentication
+
+    This API uses OAuth2 and tokens can be requested from [Dev Portal: Tokens](https://example.org/developers/tokens).
```
```suggestion
```

nulltoken replied:
@philsturgeon Fixed

philsturgeon commented on docs/reference/asyncapi-rules.md:
```diff
@@ -0,0 +1,368 @@
+
+### asyncapi-payload-unsupported-schemaFormat
+
+Spectral currently only supports payload validation against the default AsyncAPI schema.
+
+This rule will notify you when a payload is decorated with a `schemaFormat` property. Spectral won't be able to validate it and potential issues in this payload will go undetected by the `asyncapi-payload-*` rules.
+
```
```suggestion
AsyncAPI can support various `schemaFormat` values, with the default being one of the following:

```
application/vnd.aai.asyncapi;version=2.0.0
application/vnd.aai.asyncapi+json;version=2.0.0
application/vnd.aai.asyncapi+yaml;version=2.0.0
```

At this point no other schemaFormat's are supported by Spectral, so if you use them this rule will emit an `info` message and skip validating the payload. 

Other formats such as OpenAPI Schema Object, JSON Schema Draft 07 and Avro will be added in various upcoming versions. 
```

nulltoken replied:
@philsturgeon Awesome! I've slightly reworded your proposal (as we don't support setting the `schemaFormat` at all).

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,60 @@
+import { IFunction, IFunctionContext } from '../types';
+import { IFunctionResult } from '../types/function';
+
+import * as asyncApi2Schema from '../rulesets/asyncapi/schemas/schema.asyncapi2.json';
```
This could be provided in a ruleset as a function option?

nulltoken commented:
> @nulltoken with those few changes done this is all thumbs from me! When you and @P0lip are happy merge away and let's get a beta out for people to play with! 🥳

@philsturgeon Thanks a lot for the feedback.

@P0lip What's your thinking Sir?

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,78 @@
+
+import { IFunction, IFunctionResult } from '../types';
+import { getLintTargets } from '../utils';
+import { schema } from './schema';
```
We don't need to import it.

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,78 @@
+  return validator;
+};
+
+export const asyncApi2PayloadValidation: IFunction<IPayloadValidationOptions> = (
```
```suggestion
export function asyncApi2PayloadValidation = (
this: IFunctionContext,
// ..rest of args
```

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,78 @@
+  paths,
+  otherValues,
+) => {
+  const ajvValidationFn = buildAsyncApi2SchemaObjectValidator(schema);
```
```suggestion
  const ajvValidationFn = buildAsyncApi2SchemaObjectValidator(this.functions.schema);
```

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,78 @@
+  const results: IFunctionResult[] = [];
+
+  for (const relevantItem of relevantItems) {
+    const path = rootPath.slice(0);
```
How about?
```suggestion
    const path = [...rootPath, ...relevantItem.path]
```

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,78 @@
+    const path = rootPath.slice(0);
+    Array.prototype.push.apply(path, relevantItem.path);
+
+    const result = schema(
```
```suggestion
    const result = this.functions.schema(
```

P0lip commented on src/functions/asyncApi2PayloadValidation.ts:
```diff
@@ -0,0 +1,78 @@
+) => {
+  const ajvValidationFn = buildAsyncApi2SchemaObjectValidator(schema);
+
+  const relevantItems = getLintTargets(targetVal, opts.field);
```
Just for the clarity, we've discussed this one on slack - this should not be needed, since `runner` already handles it for us.

P0lip commented:
Dang, hit the button too late. :laughing:

P0lip commented on src/rulesets/reader.ts:
```diff
@@ -65,15 +65,11 @@ const createRulesetProcessor = (
-            // happens
-          }
-          return opts;
+          return { result: parse(opts.result) };
```
If I am not mistaken, we should return original options.
I suggest not touch it unless it's needed?
